### PR TITLE
[TC Automation] PTX-21440 Add test case VolAttachFAPxRestart

### PR DESCRIPTION
**Pull request by spai-px**
**What this PR does / why we need it**:
Adds test case VolAttachFAPxRestart 

**Which issue(s) this PR fixes** (optional)
Closes # PTX-21440 

**Special notes for your reviewer**:
[Jenkins run](https://jenkins.pwx.dev.purestorage.com/job/Users/job/krishna/job/tp-pxe-byoc/2680/console)
[Aetos link](https://aetos.pwx.purestorage.com/resultSet/testSetID/621355 )

### CODE REVIEW

File: tests/basic/pure_test.go [hunk 1]
```diff
@@ -2365,3 +2365,293 @@ var _ = Describe("{FADAVolMigrateValidation}", func() {
...
+		n := node.GetStorageDriverNodes()[0]
+
+		stepLog := "get the secrete of FA which is not present in pure secret"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			//get the flash array details
+			volDriverNamespace, err := Inst().V.GetVolumeDriverNamespace()
+			log.FailOnError(err, "failed to get volume driver [%s] namespace", Inst().V.String())
+
+			NonPxPureSecret, err := pureutils.GetNonPxPureSecret(volDriverNamespace)
```
[the remark this answers is not in the document]
Comment (spai-px) [reply tests/basic/pure_test.go:2406]: we need this to get secrets not present in pure.json

File: pkg/pureutils/pure_types.go [hunk 2]
```diff
@@ -76,6 +78,12 @@ func (p *PXPureSecret) GetArrayToZoneMap() map[string]string {
 	return arrayEndpointToZoneMap
 }
 
+func GetNonPxPureSecret(namespace string) (PXPureSecret, error) {
```
[the remark this answers is not in the document]
Comment (spai-px) [reply pkg/pureutils/pure_types.go:81]: So the test case is to attach a volume from FA which is not present in pure.json so I have created another file named non-pure.json which should contain the secrets not present in pure.json.

File: pkg/pureutils/pure.go [hunk 3]
```diff
@@ -156,6 +156,33 @@ func ConnectVolumeToHost(faClient *flasharray.Client, hostName string, volName s
 	return connectedVol, nil
 }
 
+// DisConnectVolumeFromHost Disconnects Volume from Host
+func DisConnectVolumeFromHost(faClient *flasharray.Client, hostName string, volName string) (*flasharray.ConnectedVolume, error) {
```
[the remark this answers is not in the document]
Comment (spai-px) [reply pkg/pureutils/pure.go:160]: Done.

File: tests/basic/pure_test.go [hunk 4]
```diff
@@ -2365,3 +2365,293 @@ var _ = Describe("{FADAVolMigrateValidation}", func() {
...
+			if err != nil {
+				return err
+			}
+			log.InfoD("Output of iscsiadm logout command: %v", iscsiAdmOutput)
+		}
+	}
+	return nil
+}
+
+func GetIQNOfNode(n node.Node) (string, error) {
```
[the remark this answers is not in the document]
Comment (spai-px) [reply tests/basic/pure_test.go:2644]: Done.

File: tests/basic/pure_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vprabhakar-px) [reply tests/basic/pure_test.go:2406]: why to get secrets , instead why not pass details from torpedo externally

File: pkg/pureutils/pure_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (spai-px) [reply pkg/pureutils/pure_types.go]: Changed it to take it from env variables.

File: tests/basic/pure_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (spai-px) [reply tests/basic/pure_test.go]: Yeah done.

File: tests/basic/pure_test.go [hunk 5]
```diff
@@ -2365,3 +2364,279 @@ var _ = Describe("{FADAVolMigrateValidation}", func() {
...
+		})
+
+		stepLog = "Create a volume, create a host, attach the volume to the host, update iqn of the host and attach the volume to the host"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			iqn, err := GetIQNOfNode(n)
+			log.FailOnError(err, "Failed to get iqn of the node %v", n.Name)
+			log.InfoD("Iqn of the node: %v", iqn)
+
+			//create a connections to the FA
```
Comment (vprabhakar-px) [inline tests/basic/pure_test.go:2441]: pls update the comment

File: tests/basic/pure_test.go [hunk 6]
```diff
@@ -2365,3 +2364,279 @@ var _ = Describe("{FADAVolMigrateValidation}", func() {
...
+			FAclient, err = pureutils.PureCreateClientAndConnect(faMgmtEndPoint, faAPIToken)
+			log.FailOnError(err, "Failed to create client and connect to FA")
+
+			//create a volume on the FA
+			volume, err := pureutils.CreateVolumeOnFABackend(FAclient, volumeName, 1048576)
```
Comment (vprabhakar-px) [inline tests/basic/pure_test.go:2446]: instead of hardcoding pick random Volume ranges from few 1GB till 100Gb

File: tests/basic/pure_test.go [hunk 7]
```diff
@@ -2365,3 +2364,279 @@ var _ = Describe("{FADAVolMigrateValidation}", func() {
...
+			log.FailOnError(err, "Failed to create volume on FA")
+			log.InfoD("Volume created on FA: %v", volume.Name)
+
+			//create a host in the FA
+			host, err := pureutils.CreateNewHostOnFA(FAclient, hostName)
+			log.FailOnError(err, "Failed to create host on FA")
+			log.InfoD("Host created on FA: %v", host.Name)
+
```
Comment (vprabhakar-px) [inline tests/basic/pure_test.go:2454]: Validate if IQN is already present on some hosts

File: tests/basic/pure_test.go [hunk 8]
```diff
@@ -2365,3 +2364,279 @@ var _ = Describe("{FADAVolMigrateValidation}", func() {
...
+	})
+
+	JustAfterEach(func() {
+		defer EndTorpedoTest()
+	})
+})
+
+func LoginIntoAllControllers(n node.Node, networkInterfaces []flasharray.NetworkInterface) error {
+	var ip string
```
Comment (vprabhakar-px) [inline tests/basic/pure_test.go:2550]: Split the function to Login to one specific controller and use the function to access and login from multiple network interfaces

**Review by vprabhakar-px — APPROVED** [review]
some changes Needed , LGTM after updating all the comments

File: tests/basic/pure_test.go (hunk 8 above)
Comment (spai-px) [reply tests/basic/pure_test.go]: Done.

File: tests/basic/pure_test.go (hunk 7 above)
Comment (spai-px) [reply tests/basic/pure_test.go]: Done. If Iqn is present taking the host which is already present.

File: tests/basic/pure_test.go (hunk 6 above)
Comment (spai-px) [reply tests/basic/pure_test.go]: Done.

File: tests/basic/pure_test.go (hunk 5 above)
Comment (spai-px) [reply tests/basic/pure_test.go]: Done.

**Review by dbhatnagar-px — APPROVED** [review]
Lgtm

**Pull request merged**